### PR TITLE
[MIST-1034] Decouple query store and state checkpointing

### LINKED ISSUE #1034: Decouple storing of query configurations and query states.
In current implementation of MIST, query configurations are periodically check-pointed with the states. However, this can be inefficient because query configurations are consistently overwritten even though they are basically immutable. To solve this problem, we need to decouple storing query configurations with the checkpointing process.

### PR DESCRIPTION
**Pull request by DifferentSC**
This PR resolves #1034 via
* Decoupling query store and state checkpointing
* Implementing necessary protocols and methods

### CODE REVIEW

File: mist-core/src/main/java/edu/snu/mist/core/task/groupaware/DefaultGroupImpl.java [hunk 1]
```diff
@@ -325,23 +325,22 @@ public GroupCheckpoint checkpoint() {
...
         .setMinimumLatestCheckpointTimestamp(groupTimestamp.getValue())
-        .setJarFilePaths(applicationInfo.getJarFilePath())
         .setGroupId(groupId)
-        .setApplicationId(applicationInfo.getApplicationId())
+        .setQueryCheckpointMap(queryCheckpointMap)
         .build();
   }
 
   /**
    * Convert a ConfigDag to an AvroConfigDag.
```
Comment (kyungissac) [inline mist-core/src/main/java/edu/snu/mist/core/task/groupaware/DefaultGroupImpl.java]: Please fix this comment.

File: mist-core/src/main/java/edu/snu/mist/core/task/QueryManager.java [hunk 2]
```diff
@@ -39,6 +40,15 @@
    */
   QueryControlResult create(AvroDag avroDag);
 
+  /**
+   * Recover a checkpointed query.
+   * @param avroDAg
```
Comment (kyungissac) [inline mist-core/src/main/java/edu/snu/mist/core/task/QueryManager.java]: Please fix this to avroDag.

File: mist-core/src/main/java/edu/snu/mist/core/task/QueryManager.java [hunk 3]
```diff
@@ -39,6 +40,15 @@
...
+   * @param checkpointedState
+   * @return
+   */
+  QueryControlResult createQueryWithCheckpoint(AvroDag avroDAg,
```
Comment (kyungissac) [inline mist-core/src/main/java/edu/snu/mist/core/task/QueryManager.java]: Please fix this to avroDag.

File: mist-core/src/main/java/edu/snu/mist/core/task/checkpointing/DefaultCheckpointManagerImpl.java [hunk 4]
```diff
@@ -189,7 +144,7 @@ public void deleteGroup(final String groupId) {
       return;
     }
     final QueryRemover remover = group.getApplicationInfo().getQueryRemover();
-    for (final Query query : group.getQueries()) {
+    for (final Query query: group.getQueries()) {
```
Comment (kyungissac) [inline mist-core/src/main/java/edu/snu/mist/core/task/checkpointing/DefaultCheckpointManagerImpl.java]: Why was this changed?

File: mist-core/src/main/java/edu/snu/mist/core/task/checkpointing/DefaultCheckpointManagerImpl.java [hunk 5]
```diff
@@ -73,97 +76,49 @@ private DefaultCheckpointManagerImpl(final ApplicationMap applicationMap,
...
+      // Load the queries.
+      queryCheckpointMap = checkpointStore.loadSavedGroupState(groupId).getQueryCheckpointMap();
+      final List<String> queryIdListInGroup = new ArrayList<>();
+      for (final String queryId : queryCheckpointMap.keySet()) {
+        queryIdListInGroup.add(queryId);
+      }
+      dagList = checkpointStore.loadSavedQueries(queryIdListInGroup);
+      // Load the states.
     } catch (final FileNotFoundException ie) {
       LOG.log(Level.WARNING, "Failed in loading app {0}, this app may not exist in the checkpoint store.",
```
Comment (kyungissac) [inline mist-core/src/main/java/edu/snu/mist/core/task/checkpointing/DefaultCheckpointManagerImpl.java]: Please also fix this log message.

File: mist-core/src/main/java/edu/snu/mist/core/task/stores/DefaultGroupCheckpointStore.java [hunk 6]
```diff
@@ -45,32 +48,76 @@
...
+   */
+  private final DatumWriter<AvroDag> avroDagDatumWriter;
+
+  /**
+   * A reader that reads stored avro dags.
+   */
+  private final DatumReader<AvroDag> avroDagDatumReader;
+
   /**
    * A writer that stores ApplicationInfoCheckpoint.
```
Comment (kyungissac) [inline mist-core/src/main/java/edu/snu/mist/core/task/stores/DefaultGroupCheckpointStore.java]: Please fix this comment.

File: mist-core/src/main/java/edu/snu/mist/core/task/stores/DefaultGroupCheckpointStore.java [hunk 7]
```diff
@@ -45,32 +48,76 @@
...
    */
-  private final DatumWriter<GroupCheckpoint> datumWriter;
+  private final DatumWriter<GroupCheckpoint> groupCheckpointDatumWriter;
 
   /**
    * A reader that reads stored ApplicationInfoCheckpoint.
```
Comment (kyungissac) [inline mist-core/src/main/java/edu/snu/mist/core/task/stores/DefaultGroupCheckpointStore.java]: Please fix this comment.

File: mist-common/src/main/avro/checkpoint_manager.avpr [hunk 8]
```diff
@@ -23,92 +23,40 @@
...
               ]
-            }
-          }
+            },
+            "null"
+          ]
         },
         {
-          "name": "AvroConfigMISTEdges",
+          "name": "LatestCheckpointTimestamp",
```
Comment (taegeonum) [inline mist-common/src/main/avro/checkpoint_manager.avpr]: `LatestCheckpointTimestamp` -> `CheckpointTimestamp`

File: mist-common/src/main/avro/checkpoint_manager.avpr [hunk 9]
```diff
@@ -119,33 +67,19 @@
...
           }
         },
         {
-          "name": "GroupId",
-          "type": "string"
-        },
-        {
-          "name": "ApplicationId",
-          "type": "string"
+          "name": "MinimumLatestCheckpointTimestamp",
```
Comment (taegeonum) [inline mist-common/src/main/avro/checkpoint_manager.avpr]: CheckpointTimestamp

File: mist-core/src/main/java/edu/snu/mist/core/task/DefaultConfigDagGeneratorImpl.java [hunk 10]
```diff
@@ -60,25 +58,50 @@ private DefaultConfigDagGeneratorImpl() {
...
+        // Create a config vertex
+        final ConfigVertex configVertex =
+            new ConfigVertex(Long.toString(configVertexId.getAndIncrement()), type, avroVertex.getConfiguration());
+        deserializedVertices.add(configVertex);
+        configDag.addVertex(configVertex);
+      }
+    } else {
+      // There are checkpointed states -> generate vertices with internal states.
+      final int numVertices = avroVertices.size();
+      final List<StateWithTimestamp> queryState = checkpointedState.getQueryState();
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/task/DefaultConfigDagGeneratorImpl.java]: Do we guarantee that the state order and vertex order are the same?

**Comment by kyungissac** [comment]
LGTM. @taegeonum do you want to take a look?

File: mist-core/src/main/java/edu/snu/mist/core/task/stores/DefaultGroupCheckpointStore.java [hunk 11]
```diff
@@ -45,32 +48,76 @@
...
   private DefaultGroupCheckpointStore(@Parameter(SharedStorePath.class) final String tmpFolderpath) {
     this.tmpFolderPath = tmpFolderpath;
-    this.datumWriter = new SpecificDatumWriter<>(GroupCheckpoint.class);
-    this.datumReader = new SpecificDatumReader<>(GroupCheckpoint.class);
+    this.avroDagDatumWriter = new SpecificDatumWriter<>(AvroDag.class);
+    this.avroDagDatumReader = new SpecificDatumReader<>(AvroDag.class);
+    this.groupCheckpointDatumWriter = new SpecificDatumWriter<>(GroupCheckpoint.class);
+    this.groupCheckpointDatumReader = new SpecificDatumReader<>(GroupCheckpoint.class);
+    // Create a folder that stores the dags and jar files
+    final File folder = new File(tmpFolderPath);
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/task/stores/DefaultGroupCheckpointStore.java]: The folder is created by master so please remove this code.

File: mist-core/src/main/java/edu/snu/mist/core/task/groupaware/GroupAwareQueryManagerImpl.java [hunk 12]
```diff
@@ -176,7 +191,12 @@ public QueryControlResult create(final AvroDag avroDag) {
       }
 
       final ApplicationInfo applicationInfo = applicationMap.get(appId);
-      final DAG<ConfigVertex, MISTEdge> configDag = configDagGenerator.generate(avroDag);
+      final DAG<ConfigVertex, MISTEdge> configDag;
+      if (checkpointedState == null) {
+        configDag = configDagGenerator.generate(avroDag);
+      } else {
+        configDag = configDagGenerator.generateWithCheckpointedStates(avroDag, checkpointedState);
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/task/groupaware/GroupAwareQueryManagerImpl.java]: Could you please explain how `PhysicalOperator` restores the state?

**Comment by DifferentSC** [comment]
@taegeonum is taking a look now.

File: mist-core/src/main/java/edu/snu/mist/core/task/groupaware/GroupAwareQueryManagerImpl.java (hunk 12 above)
Comment (DifferentSC) [reply mist-core/src/main/java/edu/snu/mist/core/task/groupaware/GroupAwareQueryManagerImpl.java]: If a `ConfigVertex` has its internal states, it is deserialized and recovered in `ExecutionDagGenerator`.

File: mist-core/src/main/java/edu/snu/mist/core/task/DefaultConfigDagGeneratorImpl.java (hunk 10 above)
Comment (DifferentSC) [reply mist-core/src/main/java/edu/snu/mist/core/task/DefaultConfigDagGeneratorImpl.java]: No. Current `DAG` interface returns the set of vertices, which means that the order of the vertices can be changed according to its implementation. We need to give the separate id for each vertex to solve this problem.

**Comment by DifferentSC** [comment]
I have addressed your comments. Thanks a lot, @taegeonum!

File: mist-core/src/main/java/edu/snu/mist/core/task/stores/DefaultGroupCheckpointStore.java [hunk 13]
```diff
@@ -46,31 +49,70 @@
...
-  public CheckpointResult saveGroupCheckpoint(final Tuple<String, Group> tuple) {
+  public boolean saveQuery(final AvroDag avroDag) {
+    final String queryId = avroDag.getQueryId();
+    try {
+      final File storedFile = getQueryStoreFile(queryId);
+      if (storedFile.exists()) {
+        storedFile.delete();
+        LOG.log(Level.INFO, "Deleting a duplicate query file");
+      }
+      storedFile.getParentFile().mkdirs();
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/task/stores/DefaultGroupCheckpointStore.java]: Why do you create directory here?

File: mist-core/src/main/java/edu/snu/mist/core/task/stores/DefaultGroupCheckpointStore.java [hunk 14]
```diff
@@ -81,14 +123,12 @@ public CheckpointResult saveGroupCheckpoint(final Tuple<String, Group> tuple) {
...
       }
-      if (!storedFile.exists()) {
-        storedFile.getParentFile().mkdirs();
-        final DataFileWriter<GroupCheckpoint> dataFileWriter = new DataFileWriter<>(datumWriter);
-        dataFileWriter.create(checkpoint.getSchema(), storedFile);
-        dataFileWriter.append(checkpoint);
-        dataFileWriter.close();
-        LOG.log(Level.INFO, "Checkpoint completed for groupId: {0}", groupId);
-      }
+      storedFile.getParentFile().mkdirs();
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/task/stores/DefaultGroupCheckpointStore.java]: Why do you create directory here?

**Comment by taegeonum** [comment]
LGTM. I'm merging it
